packages/core/bootstrap/src/lib/middleware/cache-key/index.ts

```diff
@@ -0,0 +1,46 @@
+import type {
+  Middleware,
+  AdapterRequest,
+  Config,
+  APIEndpoint,
+  InputParameters,
+} from '@chainlink/types'
+import { Validator } from '../../modules'
+import { getHashOpts, hash, excludableInternalAdapterRequestProperties } from './util'
+import crypto from 'crypto'
+import { baseInputParameters } from '../../modules/selector'
+
+const baseInputParametersCachable = Object.keys(baseInputParameters).filter(
+  (inputParam) => !excludableInternalAdapterRequestProperties.includes(inputParam),
+)
+
+export const withCacheKey: <C extends Config>(
+  endpointSelector?: (request: AdapterRequest) => APIEndpoint<C>,
+) => Middleware = (endpointSelector) => async (execute, context) => async (input: AdapterRequest) => {
+  const endpoint = endpointSelector?.(input)
+  const cacheKey =
+    endpoint && endpoint.inputParameters
+      ? getCacheKey(input, endpoint.inputParameters)
+      : hash(input, getHashOpts()) // Fallback to legacy object hash cache key
+  const inputWithCacheKey = { ...input, debug: { ...input.debug, cacheKey } }
+  return execute(inputWithCacheKey, context)
+}
+
+export function getCacheKey(request: AdapterRequest, inputParameters: InputParameters): string {
+  const inputParameterKeys = Object.keys(inputParameters ?? {}).concat(baseInputParametersCachable)
```

Comment:
an optimization here could be that since input parameters are static this could be calculated once @ bootstrap

Reply:
For `inputParameterKeys`? Or do you mean for selecting the endpoint?

Reply:
we could precompute inputParameterKeys for each adapter endpoint, so the only operation here would be getting that value from an object

Reply:
sounds good